tests/src/rules/no-anonymous-default-export.js

```diff
@@ -0,0 +1,47 @@
+import { test, SYNTAX_CASES } from '../utils'
+
+import { RuleTester } from 'eslint'
+
+var ruleTester = new RuleTester()
+var rule = require('rules/no-anonymous-default-export')
+
+ruleTester.run('no-anonymous-default-export', rule, {
+    valid: [
+      // Exports with identifiers are valid
+      test({ code: 'const foo = 123\nexport default foo' }),
+      test({ code: 'export default function foo() {}'}),
+      test({ code: 'export default class MyClass {}'}),
+
+      // Allow each forbidden type with appropriate option
+      test({ code: 'export default []', options: [{ allowArray: true }] }),
+      test({ code: 'export default () => {}', options: [{ allowArrowFunction: true }] }),
+      test({ code: 'export default class {}', options: [{ allowAnonymousClass: true }] }),
+      test({ code: 'export default function() {}', options: [{ allowAnonymousFunction: true }] }),
+      test({ code: 'export default 123', options: [{ allowLiteral: true }] }),
+      test({ code: 'export default `foo`', options: [{ allowLiteral: true }] }),
+      test({ code: 'export default {}', options: [{ allowObject: true }] }),
+
+      // Allow forbidden types with multiple options
+      test({ code: 'export default 123', options: [{ allowLiteral: true, allowObject: true }] }),
+      test({ code: 'export default {}', options: [{ allowLiteral: true, allowObject: true }] }),
+
+      // Sanity check unrelated export syntaxes
+      test({ code: 'export * from \'foo\'' }),
+      test({ code: 'const foo = 123\nexport { foo }' }),
+      test({ code: 'const foo = 123\nexport { foo as default }' }),
+
+      ...SYNTAX_CASES,
+    ],
+
+    invalid: [
+      test({ code: 'export default []', errors: [{ message: 'Unexpected default export of an array' }] }),
```

Comment:
I think the error messages are confusing.
`Unexpected default export of an anonymous class/function` are fine, because they mention that what is wrong is the anonymousness.

For the others, it feels like this rule says that arrays/literals/objects are not allowed to be the default export for some reason.
I think it would be clearer if we recommended giving it a name or assigning it to a variable.